classes/auth.php

```diff
@@ -573,42 +573,56 @@ public function saml_login_complete($attributes) {
         }
 
         $attr = $this->config->idpattr;
-        if (empty($attributes[$attr]) ) {
+        if (empty($attributes[$attr])) {
+            // Missing mapping IdP attribute. Login failed.
+            $event = \core\event\user_login_failed::create(['other' => ['username' => 'unknown',
+                'reason' => AUTH_LOGIN_NOUSER]]);
+            $event->trigger();
             $this->error_page(get_string('noattribute', 'auth_saml2', $attr));
         }
 
+        // Testing user's groups and allow access according to preferences.
```

Comment:
Shouldn't this access check happen "After" we have worked out if we have a valid user?

Reply:
It does not need user record, it checks group in the attributes against plugin configuration.

Reply:
Possibly the comment can be improved.

Reply:
of course - so in fact it makes a lot more sense to happen before we check if the user exists in Moodle - much better where you've put it there.